// nestjs-ioredis/dist/redis-core.module.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
var RedisCoreModule_1;
const common_1 = require("@nestjs/common");
const redis_provider_1 = require("./redis.provider");
const redis_constant_1 = require("./redis.constant");
let RedisCoreModule = RedisCoreModule_1 = class RedisCoreModule {
    static forRoot(options) {
        const optionProvider = this.createAsyncOptionsProvider({
            useValue: Array.isArray(options) ? options : [options],
        });
        const redisClientProviders = redis_provider_1.RedisProvider.init(this.resolveOptions(options));
        return {
            module: RedisCoreModule_1,
            providers: [optionProvider, ...redisClientProviders],
            exports: redisClientProviders,
        };
    }
    static forAsync(options, injectOption) {
        const optionProvider = this.createAsyncOptionsProvider(injectOption);
        const redisClientProviders = redis_provider_1.RedisProvider.init(this.resolveOptions(options));
        return {
            module: RedisCoreModule_1,
            providers: [optionProvider, ...redisClientProviders],
            exports: redisClientProviders,
        };
    }
    static resolveOptions(options) {
        if (!Array.isArray(options) && options.name === undefined) {
            options.name = redis_constant_1.REDIS_CLIENT_DEFAULT_KEY;
        }
        if (!Array.isArray(options)) {
            options = [options];
        }
        options.forEach((option, index) => {
            if (option.name === undefined) {
                option.name = index.toString();
            }
        });
        return options;
    }
    static createAsyncOptionsProvider(options) {
        return {
            provide: redis_constant_1.REDIS_CLIENT_MODULE_OPTIONS,
            useValue: options.useValue,
            useFactory: options.useFactory,
            inject: options.inject || [],
        };
    }
};
RedisCoreModule = RedisCoreModule_1 = __decorate([
    common_1.Global(),
    common_1.Module({})
], RedisCoreModule);
exports.RedisCoreModule = RedisCoreModule;
