// nestjs-ioredis/dist/redis.provider.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const Redis = require("ioredis");
const create_token_1 = require("./utils/create.token");
const redis_constant_1 = require("./redis.constant");
const clients = new Map();
class RedisProvider {
    static createClient(options) {
        return new Redis(options);
    }
    static init(options) {
        return options.map(option => this.createRedisClientProvider(option));
    }
    static createRedisClientProvider(option) {
        const token = create_token_1.createClientToken(option.name);
        let client;
        if (clients.get(token)) {
            client = clients.get(token);
            return {
                provide: token,
                useValue: client,
            };
        }
        else {
            return {
                provide: token,
                useFactory: (options) => {
                    const config = options.find(item => (item.name === option.name) || (option.name === redis_constant_1.REDIS_CLIENT_DEFAULT_KEY && item.name === undefined));
                    option = Object.assign({}, option, config);
                    client = this.createClient(option);
                    clients.set(token, client);
                    return client;
                },
                inject: [redis_constant_1.REDIS_CLIENT_MODULE_OPTIONS],
            };
        }
    }
}
exports.RedisProvider = RedisProvider;
